crates/moonutil/src/common.rs

```diff
@@ -48,12 +57,89 @@ pub fn startswith_and_trim(s: &str, t: &str) -> String {
     }
 }
 
-pub fn read_module_from_json(path: &Path) -> anyhow::Result<MoonMod> {
-    let file = File::open(path)?;
+#[derive(Debug)]
+pub enum SourceDirError {
+    EmptyString,
+    ContainInvalidChars(Vec<char>),
+}
+
+impl Display for SourceDirError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SourceDirError::EmptyString => write!(f, "source-dir should not be empty"),
+            SourceDirError::ContainInvalidChars(cs) => {
+                write!(f, "source-dir should not contain invalid chars `{:?}`", cs)
+            }
+        }
+    }
+}
+
+impl Error for SourceDirError {}
+
+fn is_valid_folder_name(folder_name: &str) -> Result<(), SourceDirError> {
+    if folder_name.is_empty() {
+        return Err(SourceDirError::EmptyString);
+    }
+
+    let invalid_chars = ['<', '>', ':', '"', '/', '\\', '|', '?', '*'];
+    let invalid: Vec<char> = folder_name
+        .chars()
+        .filter(|c| invalid_chars.contains(c))
+        .collect();
+    if !invalid.is_empty() {
+        return Err(SourceDirError::ContainInvalidChars(invalid));
+    }
+    Ok(())
+}
+
+#[derive(Debug)]
+pub struct MoonModJSONFormatError {
+    path: Box<Path>,
+    kind: MoonModJSONFormatErrorKind,
+}
+
+impl Display for MoonModJSONFormatError {
```

Comment:
how about use `thiserror` instead?

for example:
```rust
#[derive(Debug, Error)]
pub enum MoonModJSONFormatErrorKind {
    #[error("I/O error")]
    IO(#[from] std::io::Error),
    #[error("Parse error")]
    Parse(#[from] serde_json_lenient::Error),
    #[error("Source directory error")]
    SourceDir(#[from] SourceDirError),
}

#[derive(Debug, Error)]
#[error("failed to load `{}`", path.display())]
pub struct MoonModJSONFormatError {
    path: Box<Path>,
    #[source]
    kind: MoonModJSONFormatErrorKind,
}
```

Reply:
cool